File: src/main/java/org/lanternpowered/server/inventory/LanternItemStack.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.inventory;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.base.MoreObjects;
import org.lanternpowered.server.data.AdditionalContainerCollection;
import org.lanternpowered.server.data.DataQueries;
import org.lanternpowered.server.data.IAdditionalDataHolder;
import org.lanternpowered.server.data.IValueContainer;
import org.lanternpowered.server.data.ValueCollection;
import org.lanternpowered.server.data.property.AbstractPropertyHolder;
import org.lanternpowered.server.item.LanternItemType;
import org.spongepowered.api.data.DataContainer;
import org.spongepowered.api.data.DataView;
import org.spongepowered.api.data.key.Keys;
import org.spongepowered.api.data.manipulator.DataManipulator;
import org.spongepowered.api.data.persistence.InvalidDataException;
import org.spongepowered.api.item.ItemType;
import org.spongepowered.api.item.ItemTypes;
import org.spongepowered.api.item.inventory.ItemStack;
import org.spongepowered.api.item.inventory.ItemStackSnapshot;
import org.spongepowered.api.text.translation.Translation;

import java.util.Collections;
import java.util.HashSet;
import java.util.function.Consumer;

import javax.annotation.Nullable;

@SuppressWarnings({"ConstantConditions", "SimplifiableConditionalExpression"})
public class LanternItemStack implements ItemStack, AbstractPropertyHolder, IAdditionalDataHolder {

    private static final LanternItemStack empty = null;

    /**
     * Gets a empty {@link ItemStack} if the specified {@link ItemStack}
     * is {@code null}. Otherwise returns the item stack itself.
     *
     * @param itemStack The item stack
     * @return A empty or the provided item stack
     */
    public static LanternItemStack orEmpty(@Nullable ItemStack itemStack) {
        return itemStack == null ? empty : (LanternItemStack) itemStack;
    }

    /**
     * Gets a empty {@link ItemStack}.
     *
     * <p>A empty item stack will always have the item type
     * {@link ItemTypes#NONE} and a quantity of {@code 0}.
     * And any data offered to it will be rejected.</p>
     *
     * @return The empty item stack
     */
    public static LanternItemStack empty() {
        return empty;
    }

    private final ValueCollection valueCollection;
    private final AdditionalContainerCollection<DataManipulator<?, ?>> additionalContainers;
    private final ItemType itemType;

    private int quantity;

    /**
     * Constructs a new {@link LanternItemStack} for the specified {@link ItemType}.
     *
     * @param itemType The item type
     */
    public LanternItemStack(ItemType itemType) {
        this(itemType, 1);
    }

    /**
     * Constructs a new {@link LanternItemStack} for the specified {@link ItemType}.
     *
     * @param itemType The item type
     * @param quantity The quantity
     */
    public LanternItemStack(ItemType itemType, int quantity) {
        // Use empty containers for the none item type
        this(itemType, quantity, ValueCollection.create(),
                itemType == ItemTypes.NONE ? AdditionalContainerCollection.empty() : AdditionalContainerCollection.createConcurrent());
        registerKeys();
    }

    private LanternItemStack(ItemType itemType, int quantity, ValueCollection valueCollection,
            AdditionalContainerCollection<DataManipulator<?, ?>> additionalContainers) {
        checkArgument(quantity >= 0, "quantity may not be negative");
        checkNotNull(itemType, "itemType");
        this.additionalContainers = additionalContainers;
        this.valueCollection = valueCollection;
        this.quantity = quantity;
        this.itemType = itemType;
    }

    private void registerKeys() {
        final ValueCollection c = getValueCollection();
        ((LanternItemType) this.itemType).getKeysProvider().accept(c);
        c.register(Keys.DISPLAY_NAME, null);
        c.register(Keys.ITEM_LORE, Collections.emptyList());
        c.register(Keys.BREAKABLE_BLOCK_TYPES, new HashSet<>());
        c.register(Keys.ITEM_ENCHANTMENTS, Collections.emptyList());
    }

    @Override
    public AdditionalContainerCollection<DataManipulator<?, ?>> getAdditionalContainers() {
        return this.additionalContainers;
    }

    @Override
    public ValueCollection getValueCollection() {
        return this.valueCollection;
    }

    @Override
    public boolean validateRawData(DataView dataView) {
        return dataView.contains(DataQueries.ITEM_TYPE);
    }

    @Override
    public void setRawData(DataView dataView) throws InvalidDataException {
        checkNotNull(dataView, "dataView");
        dataView.remove(DataQueries.ITEM_TYPE);
        this.quantity = dataView.getInt(DataQueries.QUANTITY).orElse(1);
        IAdditionalDataHolder.super.setRawData(dataView);
    }

    @Override
    public DataContainer toContainer() {
        return IAdditionalDataHolder.super.toContainer()
                .set(DataQueries.ITEM_TYPE, getType())
                .set(DataQueries.QUANTITY, getQuantity());
    }

    @Override
    public Translation getTranslation() {
        return ((LanternItemType) getType()).getTranslationProvider().get(this.itemType, this);
    }

    @Override
    public ItemType getType() {
        return this.quantity == 0 ? ItemTypes.NONE : this.itemType;
    }

    @Override
    public int getQuantity() {
        return this.itemType == ItemTypes.NONE ? 0 : this.quantity;
    }

    @Override
    public void setQuantity(int quantity) throws IllegalArgumentException {
        checkArgument(quantity >= 0, "quantity may not be negative");
        this.quantity = quantity;
    }

    /**
     * Clears this {@link ItemStack}, this sets the quantity to {@code 0}.
     */
    public void clear() {
        setQuantity(0);
    }

    @Override
    public int getMaxStackQuantity() {
        return getType().getMaxStackQuantity();
    }

    @Override
    public ItemStackSnapshot createSnapshot() {
        if (isEmpty()) {
            return ItemStackSnapshot.NONE;
        }
        return new LanternItemStackSnapshot(copy());
    }

    public ItemStackSnapshot toSnapshot() {
        return createSnapshot();
    }

    public ItemStackSnapshot toWrappedSnapshot() {
        if (isEmpty() && empty != null) {
            return ItemStackSnapshot.NONE;
        }
        return new LanternItemStackSnapshot(this);
    }

    @Override
    public boolean equalTo(ItemStack that) {
        return similarTo(that) && getQuantity() == that.getQuantity();
    }

    /**
     * Similar to {@link #equalTo(ItemStack)}, but matches this
     * {@link ItemStack} with a {@link ItemStackSnapshot}.
     *
     * @param that The other snapshot
     * @return Is equal
     */
    public boolean equalTo(ItemStackSnapshot that) {
        return similarTo(that) && getQuantity() == that.getQuantity();
    }

    @Override
    public boolean isEmpty() {
        return this.itemType == ItemTypes.NONE || this.quantity <= 0;
    }

    @Override
    public LanternItemStack copy() {
        // Just return the empty instance
        if (isEmpty()) {
            return empty;
        }
        return new LanternItemStack(this.itemType, this.quantity, getValueCollection().copy(), this.additionalContainers.copy());
    }

    /**
     * Gets whether this item stack is filled. (non empty)
     *
     * @return Is filled
     */
    public boolean isFilled() {
        return !isEmpty();
    }

    /**
     * Applies the {@link Consumer} when this item stack isn't empty.
     *
     * @param consumer The consumer to accept
     * @see #isEmpty()
     */
    public void ifFilled(Consumer<LanternItemStack> consumer) {
        if (isFilled()) {
            consumer.accept(this);
        }
    }

    /**
     * Gets whether the specified {@link ItemStackSnapshot} is similar
     * to this {@link ItemStack}. The {@link ItemType} and all
     * the applied data must match.
     *
     * @param that The other snapshot
     * @return Is similar
     */
    public boolean similarTo(ItemStackSnapshot that) {
        checkNotNull(that, "that");
        return similarTo(((LanternItemStackSnapshot) that).itemStack);
    }
    /**
     *
     * Gets whether the specified {@link ItemStack} is similar
     * to this {@link ItemStack}. The {@link ItemType} and all
     * the applied data must match.
     *
     * @param that The other snapshot
     * @return Is similar
     */
    public boolean similarTo(ItemStack that) {
        checkNotNull(that, "that");
        final boolean emptyA = isEmpty();
        final boolean emptyB = that.isEmpty();
        if (emptyA != emptyB) {
            return emptyA && emptyB;
        }
        return getType() == that.getType() && IValueContainer.matchContents(this, (IValueContainer) that);
    }

    public static boolean isEmpty(@Nullable ItemStack itemStack) {
        return itemStack == null || itemStack.isEmpty();
    }

    public static boolean areSimilar(@Nullable ItemStack itemStackA, @Nullable ItemStack itemStackB) {
        return itemStackA == itemStackB ? true : itemStackA == null || itemStackB == null ? false :
                ((LanternItemStack) itemStackA).similarTo(itemStackB);
    }

    public static boolean areSimilar(@Nullable ItemStack itemStackA, @Nullable ItemStackSnapshot itemStackB) {
        return itemStackA == ((LanternItemStackSnapshot) itemStackB).itemStack ? true : itemStackA == null || itemStackB == null ? false :
                ((LanternItemStack) itemStackA).similarTo(itemStackB);
    }

    public static boolean areSimilar(@Nullable ItemStackSnapshot itemStackA, @Nullable ItemStack itemStackB) {
        return itemStackB == ((LanternItemStackSnapshot) itemStackA).itemStack ? true : itemStackA == null || itemStackB == null ? false :
                ((LanternItemStack) itemStackB).similarTo(itemStackA);
    }

    public static boolean areSimilar(@Nullable ItemStackSnapshot itemStackA, @Nullable ItemStackSnapshot itemStackB) {
        return ((LanternItemStackSnapshot) itemStackA).itemStack == ((LanternItemStackSnapshot) itemStackA).itemStack ? true :
                itemStackA == null || itemStackB == null ? false : ((LanternItemStackSnapshot) itemStackB).itemStack.similarTo(itemStackA);
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("type", getType().getKey())
                .add("quantity", getQuantity())
                .add("data", IValueContainer.valuesToString(this))
                .toString();
    }

    @Nullable
    public static LanternItemStack toNullable(@Nullable ItemStackSnapshot itemStackSnapshot) {
        return itemStackSnapshot == null || itemStackSnapshot.isEmpty() ? null :  (LanternItemStack) itemStackSnapshot.createStack();
    }

    @Nullable
    public static LanternItemStack toNullable(@Nullable ItemStack itemStack) {
        return itemStack == null || itemStack.isEmpty() ? null : (LanternItemStack) itemStack;
    }

    public static ItemStackSnapshot toSnapshot(@Nullable ItemStack itemStack) {
        return itemStack == null || itemStack.isEmpty() ? ItemStackSnapshot.NONE : itemStack.createSnapshot();
    }
}

File: src/main/kotlin/org/lanternpowered/server/inventory/LanternItemStack.kt
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.inventory

import org.lanternpowered.api.util.ToStringHelper
import org.lanternpowered.server.data.DataQueries
import org.lanternpowered.server.data.LocalDataHolderHelper
import org.lanternpowered.server.data.LocalKeyRegistry
import org.lanternpowered.server.data.LocalMutableDataHolder
import org.lanternpowered.server.data.property.PropertyHolderBase
import org.lanternpowered.server.data.value.ValueFactory
import org.lanternpowered.server.item.LanternItemType
import org.spongepowered.api.data.Keys
import org.spongepowered.api.data.persistence.DataContainer
import org.spongepowered.api.data.persistence.DataView
import org.spongepowered.api.item.ItemType
import org.spongepowered.api.item.ItemTypes
import org.spongepowered.api.item.inventory.ItemStack
import org.spongepowered.api.item.inventory.ItemStackSnapshot
import org.spongepowered.api.text.translation.Translation
import java.util.function.Consumer

class LanternItemStack private constructor(
        private val itemType: ItemType,
        private var quantity: Int,
        override val keyRegistry: LocalKeyRegistry<LanternItemStack>
) : ItemStack, PropertyHolderBase, LocalMutableDataHolder {

    /**
     * Gets whether this item stack is filled. (non empty)
     *
     * @return Is not empty
     */
    val isNotEmpty: Boolean
        get() = !isEmpty

    /**
     * Constructs a new [LanternItemStack] for the specified [ItemType].
     *
     * @param itemType The item type
     * @param quantity The quantity
     */
    @JvmOverloads
    constructor(itemType: ItemType, quantity: Int = 1) : this(itemType, quantity, LocalKeyRegistry.of()) {
        check(quantity >= 0) { "quantity may not be negative" }
        registerKeys()
    }

    private fun registerKeys() {
        keyRegistry {
            register(Keys.DISPLAY_NAME)
            register(Keys.ITEM_LORE, mutableListOf())
            register(Keys.BREAKABLE_BLOCK_TYPES, mutableSetOf())
            register(Keys.ITEM_ENCHANTMENTS, mutableListOf())
        }
    }

    override fun validateRawData(dataView: DataView): Boolean {
        return dataView.contains(DataQueries.ITEM_TYPE)
    }

    override fun setRawData(dataView: DataView) {
        dataView.remove(DataQueries.ITEM_TYPE)
        this.quantity = dataView.getInt(DataQueries.QUANTITY).orElse(1)
        super.setRawData(dataView)
    }

    override fun toContainer(): DataContainer = super.toContainer()
            .set(DataQueries.ITEM_TYPE, this.type)
            .set(DataQueries.QUANTITY, this.quantity)

    override fun getTranslation(): Translation {
        return (type as LanternItemType).translationProvider.get(this.itemType, this)
    }

    override fun getType(): ItemType = if (this.quantity == 0) ItemTypes.AIR else this.itemType

    override fun getMaxStackQuantity(): Int = this.type.maxStackQuantity
    override fun getQuantity(): Int = if (this.itemType === ItemTypes.AIR) 0 else this.quantity

    override fun setQuantity(quantity: Int) {
        check(quantity >= 0) { "quantity may not be negative" }
        this.quantity = quantity
    }

    override fun createSnapshot(): ItemStackSnapshot {
        return if (isEmpty) {
            ItemStackSnapshot.empty()
        } else LanternItemStackSnapshot(copy())
    }

    fun toSnapshot(): ItemStackSnapshot = createSnapshot()

    fun toWrappedSnapshot(): ItemStackSnapshot {
        return if (isEmpty && empty != null) {
            ItemStackSnapshot.empty()
        } else LanternItemStackSnapshot(this)
    }

    override fun equalTo(that: ItemStack): Boolean {
        return similarTo(that) && getQuantity() == that.quantity
    }

    /**
     * Similar to [.equalTo], but matches this
     * [ItemStack] with a [ItemStackSnapshot].
     *
     * @param that The other snapshot
     * @return Is equal
     */
    fun equalTo(that: ItemStackSnapshot): Boolean {
        return similarTo(that) && getQuantity() == that.quantity
    }

    override fun isEmpty(): Boolean {
        return this.itemType === ItemTypes.AIR || this.quantity <= 0
    }

    override fun copy(): LanternItemStack {
        // Just return the empty instance
        return if (isEmpty) {
            empty
        } else LanternItemStack(this.itemType, this.quantity, keyRegistry.copy())
    }

    /**
     * Applies the [Consumer] when this item stack isn't empty.
     *
     * @param consumer The consumer to accept
     * @see .isEmpty
     */
    fun ifNotEmpty(consumer: Consumer<LanternItemStack>) {
        if (!isEmpty) {
            consumer.accept(this)
        }
    }

    /**
     * Gets whether the specified [ItemStackSnapshot] is similar
     * to this [ItemStack]. The [ItemType] and all
     * the applied data must match.
     *
     * @param that The other snapshot
     * @return Is similar
     */
    fun similarTo(that: ItemStackSnapshot): Boolean {
        return similarTo((that as LanternItemStackSnapshot).unwrap())
    }

    /**
     *
     * Gets whether the specified [ItemStack] is similar
     * to this [ItemStack]. The [ItemType] and all
     * the applied data must match.
     *
     * @param that The other snapshot
     * @return Is similar
     */
    fun similarTo(that: ItemStack): Boolean {
        val emptyA = isEmpty
        val emptyB = that.isEmpty
        return if (emptyA != emptyB) {
            emptyA && emptyB
        } else type === that.type && LocalDataHolderHelper.matchContents(this, that as LanternItemStack)
    }

    override fun toString() = ToStringHelper(this)
            .add("type", this.type.key)
            .add("quantity", this.quantity)
            .add("data", ValueFactory.toString(this))
            .toString()

    companion object {

        private val empty by lazy { LanternItemStack(ItemTypes.AIR) }

        /**
         * Gets a empty [ItemStack] if the specified [ItemStack]
         * is `null`. Otherwise returns the item stack itself.
         *
         * @param itemStack The item stack
         * @return A empty or the provided item stack
         */
        @JvmStatic
        fun orEmpty(itemStack: ItemStack?): LanternItemStack
            = if (itemStack == null) this.empty else itemStack as LanternItemStack

        /**
         * Gets a empty [ItemStack].
         *
         * A empty item stack will always have the item type
         * [ItemTypes.AIR] and a quantity of `0`.
         * And any data offered to it will be rejected.
         *
         * @return The empty item stack
         */
        @JvmStatic
        fun empty(): LanternItemStack
            = this.empty

        @JvmStatic
        fun isEmpty(itemStack: ItemStack?): Boolean {
            return itemStack == null || itemStack.isEmpty
        }

        @JvmStatic
        fun areSimilar(itemStackA: ItemStack?, itemStackB: ItemStack?): Boolean {
            return if (itemStackA === itemStackB)
                true
            else if (itemStackA == null || itemStackB == null)
                false
            else
                (itemStackA as LanternItemStack).similarTo(itemStackB)
        }

        @JvmStatic
        fun areSimilar(itemStackA: ItemStack?, itemStackB: ItemStackSnapshot?): Boolean {
            return if (itemStackA === (itemStackB as LanternItemStackSnapshot).unwrap())
                true
            else if (itemStackA == null || itemStackB == null)
                false
            else
                (itemStackA as LanternItemStack).similarTo(itemStackB)
        }

        @JvmStatic
        fun areSimilar(itemStackA: ItemStackSnapshot?, itemStackB: ItemStack?): Boolean {
            return if (itemStackB === (itemStackA as LanternItemStackSnapshot).unwrap())
                true
            else if (itemStackA == null || itemStackB == null)
                false
            else
                (itemStackB as LanternItemStack).similarTo(itemStackA)
        }

        @JvmStatic
        fun areSimilar(itemStackA: ItemStackSnapshot?, itemStackB: ItemStackSnapshot?): Boolean {
            return if ((itemStackA as LanternItemStackSnapshot).unwrap() === (itemStackA as LanternItemStackSnapshot).unwrap())
                true
            else if (itemStackA == null || itemStackB == null) false else (itemStackB as LanternItemStackSnapshot).unwrap().similarTo(itemStackA)
        }

        @JvmStatic
        fun toNullable(itemStackSnapshot: ItemStackSnapshot?): LanternItemStack? {
            return if (itemStackSnapshot == null || itemStackSnapshot.isEmpty) null else itemStackSnapshot.createStack() as LanternItemStack
        }

        @JvmStatic
        fun toNullable(itemStack: ItemStack?): LanternItemStack? {
            return if (itemStack == null || itemStack.isEmpty) null else itemStack as LanternItemStack?
        }

        @JvmStatic
        fun toSnapshot(itemStack: ItemStack?): ItemStackSnapshot {
            return if (itemStack == null || itemStack.isEmpty) ItemStackSnapshot.empty() else itemStack.createSnapshot()
        }
    }
}
